core/arch/arm/sta/sta_debug.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2015, STMicroelectronics International N.V.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <compiler.h>
+#include <stdio.h>
+#include <trace.h>
+#include <user_ta_header.h>
+#include <mm/tee_pager.h>
+#include <string.h>
+
+#define TA_NAME		"sta_debug.ta"
+
+#define STA_DEBUG_UUID \
+		{ 0xd96a5b40, 0xe2c7, 0xb1af, \
+			{ 0x87, 0x94, 0x10, 0x02, 0xa5, 0xd5, 0xc6, 0x1b } }
+
+#define STA_DEBUG_CMD_CORE_STATUS	0
+
+#define UPDATE_BUFFER_START(buf, rem, incr, incr_all, ret_value)	\
+	do {								\
+		(incr_all) += (incr);					\
+		if ((incr) >= (rem)) return (ret_value);		\
+		(buf) += (incr);					\
+		(rem) -= (incr);					\
+	} while (0)
+
+#ifdef CFG_WITH_PAGER
+#include <mm/tee_pager.h>
+static size_t log_pager_status(char *buf, size_t rem)
+{
+	size_t l;
+	size_t all = 0;
+
+	l = snprintf(buf, rem, "Pager pool: %d / %d\n",
+		     _tee_pager_stats.npages, _tee_pager_stats.npages_all);
```

Comment:
There are other statistics ready here, such as the one on the allocation, ta loaded,... In our current implementation, everything is output in the same buffer as a string, using the same command TA.

But then I guess we could refactor this current code to have different commands, outputing only the required statistics (1 command for the pager, 1 for the allocation,...), and returning the values instead of a string as you pointed out.

I will refactor this code as you suggested.